// locklift/internal/config/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadConfig = exports.JoiConfig = exports.ConfigState = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const commander_1 = __importDefault(require("commander"));
const nodejs_1 = require("everscale-standalone-client/nodejs");
const everscale_crypto_1 = require("everscale-crypto");
const joi_1 = __importDefault(require("joi"));
var ConfigState;
(function (ConfigState) {
    ConfigState[ConfigState["EXTERNAL"] = 0] = "EXTERNAL";
    ConfigState[ConfigState["INTERNAL"] = 1] = "INTERNAL";
})(ConfigState = exports.ConfigState || (exports.ConfigState = {}));
exports.JoiConfig = joi_1.default.object({
    compiler: joi_1.default.alternatives([
        joi_1.default.object({
            includesPath: joi_1.default.string().optional(),
            externalContracts: joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.array().items(joi_1.default.string())),
            path: joi_1.default.string(),
        }),
        joi_1.default.object({
            includesPath: joi_1.default.string().optional(),
            externalContracts: joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.array().items(joi_1.default.string())),
            version: joi_1.default.string(),
        }),
    ]),
    linker: joi_1.default.alternatives([
        joi_1.default.object({
            path: joi_1.default.string(),
            lib: joi_1.default.string(),
        }),
        joi_1.default.object({
            version: joi_1.default.string(),
        }),
    ]),
    networks: joi_1.default.object().pattern(joi_1.default.string(), joi_1.default.object({
        giver: joi_1.default.alternatives().conditional(joi_1.default.object({ phrase: joi_1.default.string().required() }).unknown(), {
            then: joi_1.default.object({
                address: joi_1.default.string(),
                giverFactory: joi_1.default.any().optional(),
                phrase: joi_1.default.string(),
                accountId: joi_1.default.number(),
            }),
            otherwise: joi_1.default.object({
                address: joi_1.default.string(),
                giverFactory: joi_1.default.any().optional(),
                key: joi_1.default.string(),
            }),
        }),
        keys: joi_1.default.object({
            path: joi_1.default.string().optional(),
            phrase: joi_1.default.string().optional(),
            amount: joi_1.default.number(),
        }),
        connection: joi_1.default.alternatives([
            ...Object.keys(nodejs_1.NETWORK_PRESETS).map(el => el),
            joi_1.default.object({
                id: joi_1.default.number().optional(),
                type: joi_1.default.alternatives(["graphql", "jrpc"]),
                group: joi_1.default.string().optional(),
                data: joi_1.default.alternatives().conditional("type", {
                    is: "graphql",
                    then: joi_1.default.object({
                        endpoints: joi_1.default.array().items(joi_1.default.string()),
                        local: joi_1.default.boolean().optional(),
                        latencyDetectionInterval: joi_1.default.number().optional(),
                        maxLatency: joi_1.default.number().optional(),
                    }),
                    otherwise: joi_1.default.object({
                        endpoint: joi_1.default.string(),
                    }),
                }),
            }),
        ]),
        tracing: joi_1.default.object({
            endpoint: joi_1.default.string(),
        }).optional(),
    }).unknown()),
    mocha: joi_1.default.object({
        tsconfig: joi_1.default.string().optional(),
    }).unknown(true),
}).unknown();
function loadConfig(configPath) {
    const resolvedConfigPath = path_1.default.resolve(process.cwd(), configPath);
    if (!fs_1.default.existsSync(resolvedConfigPath)) {
        throw new commander_1.default.InvalidOptionArgumentError(`Config at ${configPath} not found!`);
    }
    const configFile = require(resolvedConfigPath);
    const validationResult = exports.JoiConfig.validate(configFile.default);
    if (validationResult.error) {
        console.log(validationResult.error.annotate());
        throw new Error("");
    }
    const config = validationResult.value;
    for (const value of Object.values(config.networks)) {
        if (value.keys != null) {
            value.keys = {
                ...value.keys,
                phrase: value.keys?.phrase || (0, everscale_crypto_1.generateBip39Phrase)(12),
                path: value.keys.path || "m/44'/396'/0'/0/INDEX",
            };
        }
    }
    return config;
}
exports.loadConfig = loadConfig;
