// everscale-standalone-client/client/ConnectionController/jrpc.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.JrpcSocket = void 0;
const core_1 = __importDefault(require("../../core"));
const { nekoton, fetch, fetchAgent } = core_1.default;
class JrpcSocket {
    async connect(clock, params) {
        class JrpcSender {
            constructor(params) {
                this.endpoint = params.endpoint;
                this.endpointAgent = fetchAgent(this.endpoint);
                this.alternativeEndpoint = params.alternativeEndpoint != null ? params.alternativeEndpoint : params.endpoint;
                this.alternativeEndpointAgent = fetchAgent(this.alternativeEndpoint);
            }
            send(data, handler, requiresDb) {
                (async () => {
                    try {
                        const url = requiresDb ? this.endpoint : this.alternativeEndpoint;
                        const agent = requiresDb ? this.endpointAgent : this.alternativeEndpointAgent;
                        const response = await fetch(url, {
                            method: 'post',
                            headers: DEFAULT_HEADERS,
                            body: data,
                            agent,
                        }).then(response => response.text());
                        handler.onReceive(response);
                    }
                    catch (e) {
                        handler.onError(e);
                    }
                })();
            }
        }
        return new nekoton.JrpcConnection(clock, new JrpcSender(params));
    }
}
exports.JrpcSocket = JrpcSocket;
const DEFAULT_HEADERS = {
    'Content-Type': 'application/json',
};
